Add error to Attempt in useAsync

ravicious opened the pull request:
## Motivation

The master version of useAsync preserves only the message property (`statusText`) of any error thrown during the action. It is good enough for 90% of use cases where you only want to show the error in the UI and do nothing else.

However, sometimes you want to show different UI [depending on what kind of error was thrown](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error#handling_a_specific_error_type). The best way to do this is to use [a custom error class](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error#custom_error_types), rather than depending on the error message. But since `useAsync` preserves only the message property through `statusText`, this is not possible. This leads to workarounds which make the code more complicated than it could be, see https://github.com/gravitational/teleport/pull/30910#discussion_r1304229057.

Another benefit of this approach, and actually the main reason I implemented this, is that custom errors can carry custom fields, acting almost like enum variants with extra data. Here's [an example](https://github.com/gravitational/teleport/pull/30910/commits/dee0ca4d54649bf4d2bcd906f3f343af75c35a0b).

Hence this PR. It adds `error` property to the `error` variant of `Attempt`. It keeps `statusText` on other variants for compatibility purposes. There's about 15 instances, mostly in tests, where `statusText` is read without inspecting `status` first.

Otherwise I tried to keep the behavior of `useAsync` the same to guarantee that everything else keeps working as usual.

## Alternatives

There's not much else we could do instead. In other languages such as Elm, our `Attempt<T>` is represented as `Attempt<Data, Error>` where `Error` is an enum describing possible errors that the given function can return (see [RemoteData](https://package.elm-lang.org/packages/krisajenkins/remotedata/latest/RemoteData#RemoteData)). In JS that is not possible, because any function can throw anything at any point, it doesn't even have to be an instance of `Error`. This is also why `Promise` has the signature of `Promise<T>` and not `Promise<T, Err>`, because it's impossible to define what `Err` is (other than `any`).

## Concerns

When the idea first came to my mind, [I was concerned with storing non-serializable data in React state](https://github.com/gravitational/teleport/pull/30910#discussion_r1303971238). It turns out that it's not really a concern, it's a totally feasible thing to do, as the comment explains.

The only caveat would be if we wanted to store the state of `useAsync` on disk, as we do with some other pieces of data in Connect. However, as I also explained in that comment, `useAsync` usually represents the result of a transient operation (fetching someone from somewhere). Typically there's no need to store the result of such operation on disk and so far we don't have any instances of this in the codebase.

ravicious committed:
Add error to Attempt in useAsync

ravicious commented on web/packages/shared/hooks/useAsync.ts:
```diff
@@ -93,8 +93,9 @@ export function useAsync<Args extends unknown[], AttemptData>(
-        ...prevState,
         status: 'processing',
+        data: prevState['data'],
+        statusText: prevState['statusText'],
```
Here I wanted to remove spread in favor of explicitly setting the fields. This shouldn't change the behavior as `Attempt` used to have those three fields only.

ravicious commented on web/packages/shared/hooks/useAsync.ts:
```diff
@@ -125,10 +126,10 @@ export function useAsync<Args extends unknown[], AttemptData>(
-            statusText: err?.message,
+            error: err,
+            statusText: err && err['message'],
             data: null,
```
This spread here was unnecessary as we overwrote all three possible fields anyway.

ravicious commented on web/packages/shared/hooks/useAsync.ts:
```diff
@@ -163,13 +164,43 @@ export class CanceledError extends Error {
-  status: AttemptStatus;
-  statusText: string;
-};
+export type Attempt<T> =
+  | {
+      status: '';
+      data: null;
+      /**
+       * @deprecated statusText is present for compatibility purposes only. To use statusText, check
+       * if status equals 'error' first.
+       */
+      statusText: string;
+    }
+  | {
+      status: 'processing';
+      /** data is either null or contains data from the previous success attempt if the attempt was retried. */
+      data: null | T;
+      /**
+       * @deprecated statusText is present for compatibility purposes only. To use statusText, check
+       * if status equals 'error' first.
+       */
+      statusText: string;
+    }
+  | {
+      status: 'success';
+      data: T;
+      /**
+       * @deprecated statusText is present for compatibility purposes only. To use statusText, check
+       * if status equals 'error' first.
+       */
+      statusText: string;
+    }
+  | {
+      status: 'error';
+      data: null;
+      statusText: string;
+      error: any;
+    };
```
Representing `Attempt` like this helps with understanding which fields are available when. `processing` shouldn't have access to the `error` field and so on.

avatus replied:
I think this is good. I was trying to think of a scenario where we'd want to know the previous error while processing another request but now that I think about it, errors are always removed when trying the next request as to not confuse stale errors so, yeah this is great!

ravicious replied:
Regarding stale values, in most implementations of a similar data structure that I'm familiar with (bringing up RemoteData from Elm again as the golden standard), it's not even possible to read stale _data_ when the action is being retried and is in progress.

In our implementation, this is totally possible as `data` gets overriden only on success or error.

---

Your comment also made me realize that we override `error` only if the callback rejects, so in theory it's possible to see something like `{status: 'success', data: /* something */, error: Error, statusText: 'some error message'}`. However, TS would not allow you to access `error` in that case anyway. And changing this behavior could technically introduce a regression so I guess I'm not going to touch it for now.

avatus commented:
Was out most of Friday but I'm VERY excited about this. Will review on Monday asap

gzdunek commented on web/packages/shared/hooks/useAsync.test.ts:
```diff
@@ -171,3 +171,26 @@ test('run does not update state after being re-run when the callback returns a r
+  await waitFor(() => result.current[0].status === 'error');
+
+  const attempt = result.current[0];
+  expect(attempt['error']).toEqual(expectedError);
```
`toEqual` just compares recursively all properties of objects, so in our case it will only compare the error message, leaving the object type unchecked.

I think it would be better to just compare references of the objects: `(attempt['error'] === expectedError).toEqual(true)`. I wanted to propose using `toBe/toStrictEqual`, but then I read [this article](https://dev.to/thejaredwilcurt/why-you-should-never-use-tobe-in-jest-48ca).

ravicious replied:
I think the article is a little bit too overzealous. In most dynamic languages I used, there is a distinction on the language level between two "entities" being comparable to each other vs two "entities" pointing to the same object in memory.

In 90% of cases I typically use the former, but the language and the testing frameworks typically have something for the latter as well. So I think `toBe` is completely fine to make sure that `attempt.error` is the same error that was returned from the callback.

And `toBe` also gives a nice error message if the error serializes to the same string but happens to refer to a different object in memory:

```
error and statusText are set when the callback returns a rejected promise

    expect(received).toBe(expected) // Object.is equality

    If it should pass with deep equality, replace "toBe" with "toStrictEqual"

    Expected: [Error: whoops]
    Received: serializes to the same string
```

gzdunek commented on web/packages/teleterm/src/ui/ConnectMyComputer/connectMyComputerContext.test.tsx:
```diff
@@ -117,7 +117,7 @@ test('startAgent re-throws errors that are thrown while spawning the process', a
   expect(result.current.currentAction).toStrictEqual({
     kind: 'start',
-    attempt: makeErrorAttempt(AgentProcessError.name),
+    attempt: makeErrorAttempt(new AgentProcessError()),
```
In production code (`currentAction.attempt.statusText !== AgentProcessError.name`) we compare an error using `.name` property. Do we want to refactor it here?

gzdunek reviewed (commented):
Thanks for adding this!
I also like that we will have better types for `Attempt<T>`.

ravicious committed:
Ensure attempt.error refers to the same object as expectedError

ravicious committed:
Refactor error handling to use instanceof with attempt.error

ravicious committed:
makeErrorAttempt: Require an Error object

In real code, the function can throw any object. But since makeErrorAttempt
is a helper function, we can make it require an Error object.

avatus commented:
Really stoked for this! I have a few places I can refactor once this goes in. Thank you

ravicious commented:
> I have a few places I can refactor once this goes in.

Keep it mind that this is only `useAsync`, but our other hooks could probably be refactored to also include an error. Though I haven't looked into it.
